File: bdk-jvm/lib/src/test/kotlin/org/bitcoindevkit/LiveTransactionTests.kt
package org.bitcoindevkit

import kotlin.test.Test

private const val SIGNET_ESPLORA_URL = "http://signet.bitcoindevkit.net"
private const val TESTNET_ESPLORA_URL = "https://esplora.testnet.kuutamo.cloud"

class LiveTransactionTests {
    @Test
    fun testSyncedBalance() {
        val descriptor: Descriptor = Descriptor(
            "wpkh(tprv8ZgxMBicQKsPf2qfrEygW6fdYseJDDrVnDv26PH5BHdvSuG6ecCbHqLVof9yZcMoM31z9ur3tTYbSnr1WBqbGX97CbXcmp5H6qeMpyvx35B/84h/1h/0h/0/*)",
            Network.SIGNET
        )
        val wallet: Wallet = Wallet.newNoPersist(descriptor, null, Network.SIGNET)
        val esploraClient: EsploraClient = EsploraClient(SIGNET_ESPLORA_URL)
        val fullScanRequest: FullScanRequest = wallet.startFullScan()
        val update = esploraClient.fullScan(fullScanRequest, 10uL, 1uL)
        wallet.applyUpdate(update)
        wallet.commit()
        println("Wallet balance: ${wallet.getBalance().total.toSat()}")

        assert(wallet.getBalance().total.toSat() > 0uL) {
            "Wallet balance must be greater than 0! Please send funds to ${wallet.revealNextAddress(KeychainKind.EXTERNAL).address.asString()} and try again."
        }

        val transaction: Transaction = wallet.transactions().first().transaction
        println("First transaction:")
        println("Txid: ${transaction.txid()}")
        println("Version: ${transaction.version()}")
        println("Total size: ${transaction.totalSize()}")
        println("Vsize: ${transaction.vsize()}")
        println("Weight: ${transaction.weight()}")
        println("Coinbase transaction: ${transaction.isCoinbase()}")
        println("Is explicitly RBF: ${transaction.isExplicitlyRbf()}")
        println("Inputs: ${transaction.input()}")
        println("Outputs: ${transaction.output()}")
    }
}

File: bdk-swift/Tests/BitcoinDevKitTests/LiveTransactionTests.swift
import XCTest
@testable import BitcoinDevKit

private let SIGNET_ESPLORA_URL = "https://blockstream.info/signet/api/"
private let TESTNET_ESPLORA_URL = "https://esplora.testnet.kuutamo.cloud"


final class LiveTransactionTests: XCTestCase {
    private let descriptor = try! Descriptor(
    descriptor: "wpkh(tprv8ZgxMBicQKsPf2qfrEygW6fdYseJDDrVnDv26PH5BHdvSuG6ecCbHqLVof9yZcMoM31z9ur3tTYbSnr1WBqbGX97CbXcmp5H6qeMpyvx35B/84h/1h/1h/0/*)", 
    network: Network.signet
    )
    private let changeDescriptor = try! Descriptor(
        descriptor: "wpkh(tprv8ZgxMBicQKsPf2qfrEygW6fdYseJDDrVnDv26PH5BHdvSuG6ecCbHqLVof9yZcMoM31z9ur3tTYbSnr1WBqbGX97CbXcmp5H6qeMpyvx35B/84h/1h/1h/1/*)", 
        network: Network.signet
    )

    func testSyncedBalance() throws {
        let persister = try Persister.newInMemory()
        let wallet = try Wallet(
            descriptor: descriptor,
            changeDescriptor: changeDescriptor,
            network: .signet,
            persister: persister
        )
        let esploraClient = EsploraClient(url: SIGNET_ESPLORA_URL)
        let fullScanRequest: FullScanRequest = try wallet.startFullScan().build()
        let update = try esploraClient.fullScan(
            request: fullScanRequest,
            stopGap: 10,
            parallelRequests: 1
        )
        try wallet.applyUpdate(update: update)
        let address = wallet.revealNextAddress(keychain: KeychainKind.external).address.description

        XCTAssertGreaterThan(
            wallet.balance().total.toSat(),
            UInt64(0),
            "Wallet must have positive balance, please send funds to \(address)"
        )
            
        guard let transaction = wallet.transactions().first?.transaction else {
            print("No transactions found")
            return
        }
        print("First transaction:")
        print("Txid: \(transaction.computeTxid())")
        print("Version: \(transaction.version())")
        print("Total size: \(transaction.totalSize())")
        print("Vsize: \(transaction.vsize())")
        print("Weight: \(transaction.weight())")
        print("Coinbase transaction: \(transaction.isCoinbase())")
        print("Is explicitly RBF: \(transaction.isExplicitlyRbf())")
        print("Inputs: \(transaction.input())")
        print("Outputs: \(transaction.output())")

    }
}
